pintc/src/asm_gen.rs

```diff
@@ -103,6 +104,28 @@ pub fn compile_contract(
         }
     }
 
+    // check for duplicate predicates
+    let mut unique_addresses: FxHashSet<&ContentAddress> = FxHashSet::default();
+    let mut original_predicate_names: FxHashMap<&ContentAddress, &String> = FxHashMap::default();
+    for (string, (_, content_address)) in &compiled_predicates {
+        // TODO-ian: find a way to keep track of span
+        // if needed, can rely on order because compiled_predicates is a FxHashMap
+
+        if !unique_addresses.insert(content_address) {
+            let original_name = original_predicate_names.get(content_address).unwrap();
+
+            handler.emit_err(Error::Compile {
+                error: CompileError::IdenticalPredicates {
+                    original_name: original_name.to_string(),
+                    duplicate_name: string.to_string(),
+                    span: empty_span(),
+                },
+            });
+        } else {
+            original_predicate_names.insert(content_address, string);
+        }
+    }
+
```

Comment:
@mohammadfawaz @otrho Do ya'll know if we have easy access to the predicate's name span? 

Right now the error produced is

```
predicate Foo() {
}

predicate Bar() {
}

predicate Beast(x: int) {
    constraint x > 2;
}

predicate Animal(x: int) {
    constraint x > 2;
}

results in 

Error: Identical predicates found in the same contract
   ╭─[:1:1]
   │
 1 │ <none>
   │ │ 
   │ ╰─ predicates `::Beast` and `::Animal` have identical constraints
   │ 
   │ Note: contract cannot compile unless each predicate has a unique overall set of constraints, even if individual constraints within the set are not unique
───╯
Error: Identical predicates found in the same contract
   ╭─[:1:1]
   │
 1 │ <none>
   │ │ 
   │ ╰─ predicates `::Bar` and `::Foo` have identical constraints
   │ 
   │ Note: contract cannot compile unless each predicate has a unique overall set of constraints, even if individual constraints within the set are not unique
───╯
```

Looking at `predicate` and `compiled_predicate` we don't appear to keep track.

I'll also take a look to see if we have any errors that don't require a span and follow suit temporarily.

Reply:
We keep track of the `span` for any `param` for the `predicate`. That's the closest I've seen to the location of the `name` of the `predicate`.

Reply:
Or if not we could just stuff the predicate name span into the predicate from the parser.

Reply:
I had to make `symbols` public

https://github.com/essential-contributions/pint/blob/6aa8453946d8acba661d25433fd0889c932bb6a3/pintc/src/predicate.rs#L988-L991

If that's not alright, then I can do as @otrho suggests and add the info to our `predicate` struct